packages/devtools-connect/src/fast-failure-connect.ts

```diff
@@ -1,21 +1,48 @@
 // It probably makes sense to put this into its own package/repository once
 // other tools start using it.
 
+const NODE_SOCKET_NON_RETRY_CODES = [
+  'ECONNREFUSED',
+  'ENOTFOUND',
+  'ENETUNREACH',
+  'EINVAL',
+];
+const COMPASS_SOCKET_SERVICE_NON_RETRY_CODES = [
+  3000, // UNAUTHORIZED
+  3003, // FORBIDDEN
+  4004, // NOT_FOUND
+  1008, // VIOLATED_POLICY
+  4101, // DO_NOT_TRY_AGAIN
+];
+
+const isCompassSocketServiceError = handleNestedErrors(
+  (error: Error & { code?: string | number }): boolean => {
+    if (error.name === 'CompassSocketServiceError') {
+      return (
+        typeof error.code === 'number' &&
+        COMPASS_SOCKET_SERVICE_NON_RETRY_CODES.includes(error.code)
+      );
+    }
+    return false;
+  },
+);
+
 function isFastFailureConnectionSingleError(
-  error: Error & { code?: string },
+  error: Error & { code?: string | number },
 ): boolean {
   switch (error.name) {
     case 'MongoNetworkError':
-      return /\b(ECONNREFUSED|ENOTFOUND|ENETUNREACH|EINVAL)\b/.test(
-        error.message,
-      );
+      return new RegExp(
+        String.raw`\b(${NODE_SOCKET_NON_RETRY_CODES.join('|')})\b`,
```

Comment:
def overkill for me to hoist the node error codes but figured the pattern matched my new front end code stuff... this stuff is very unchanging so we can go back to _not_ changing it. 😅